fetch/fetch.go

```diff
@@ -479,26 +507,115 @@ func (f *Fetch) requestHashBatchFromPeers() {
 	}
 	f.activeReqM.Unlock()
 
-	// send in batches
-	for i := 0; i < len(requestList); i += f.cfg.BatchSize {
-		j := i + f.cfg.BatchSize
-		if j > len(requestList) {
-			j = len(requestList)
+	f.send(requestList)
+}
+
+func (f *Fetch) send(requests []requestMessage) {
+	if len(requests) == 0 {
+		return
+	}
+	if f.stopped() {
+		return
+	}
+
+	retries := 0
+	for {
+		peer2batches := f.organizeRequests(requests)
+		requests = requests[:0]
+
+		for peer, peerBatches := range peer2batches {
+			for _, reqs := range peerBatches {
+				err := f.sendBatch(peer, reqs)
+				if err != nil {
+					requests = append(requests, reqs...)
+				}
+			}
+		}
+
+		if len(requests) == 0 {
+			break
 		}
-		f.sendBatch(requestList[i:j])
+
+		retries++
+		if retries > f.cfg.MaxRetriesForBatch {
+			break
+		}
+	}
+}
+
+func peerMapToList(m map[p2p.Peer]struct{}) []p2p.Peer {
+	result := make([]p2p.Peer, 0, len(m))
+	for k := range m {
+		result = append(result, k)
 	}
+	return result
 }
 
-// sendBatch dispatches batched request messages.
-func (f *Fetch) sendBatch(requests []requestMessage) {
+func (f *Fetch) organizeRequests(requests []requestMessage) map[p2p.Peer][][]requestMessage {
+	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
+	peer2requests := make(map[p2p.Peer][]requestMessage)
+
+	// max retries to find another peer with fewer requests attached
+	const maxRetries = 3
+
+	for _, req := range requests {
+		hashPeersMap, exists := f.hashToPeers[req.Hash]
+		var hashPeers []p2p.Peer
+		if exists {
+			f.HashToPeersCacheHit()
+			hashPeers = peerMapToList(hashPeersMap.peers)
+		}
+
+		f.HashToPeersCacheMiss()
+
+		var p p2p.Peer
+		if len(hashPeers) > 0 {
+			p = hashPeers[rng.Intn(len(hashPeers))]
+		} else {
+			p = GetRandomPeer(f.net.GetPeers())
+		}
```

Comment:
you should be able to combine the two if/else blocks above

Reply:
good point, fixed, cheers